---
title: "single-cell RNA-seq analysis"
output: html_document
---

```{r global-options, include=FALSE}
knitr::opts_chunk$set( fig.width=16, fig.height=8, echo=FALSE, warning=FALSE, message=FALSE)
```

Dataset: Peripheral Blood Mononuclear Cells (PBMC) from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500.
Code is updated from Seurat package https://satijalab.org/seurat/

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
```

Read10X function reads in the output of the [cellranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) pipeline from 10X Genomics, returning a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).

```{r}
pbmcdf.data <- Read10X("filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmcdf <- CreateSeuratObject(counts = pbmcdf.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmcdf

```
Check the count matrix

```{r}
# check a few genes in the first 25 cells
pbmcdf.data[c("LYZ", "IGLL5", "GNLY", "FTL"), 1:25]

dense.size <- object.size(as.matrix(pbmcdf.data))
cat('Dense size is:', dense.size)

sparse.size <- object.size(pbmcdf.data)
cat('Sparse size is:', sparse.size)
cat('Ration of Dense/Sparse is:', dense.size/sparse.size)

```
 Most values in an scRNA-seq matrix are 0, Seurat uses sparse-matrix to save memory and speed 

# pre-processing
Includes the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable features.

### QC and selecting cells for further analysis
```{r}
# Add columns to object metadata to stash QC stats
pbmcdf[["percent.mt"]] <- PercentageFeatureSet(pbmcdf, pattern = "^MT-")
# check QC metrics for the first a few cells
head(pbmcdf@meta.data, 10)
# Visualize by violin plot
VlnPlot(pbmcdf, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# Scatter plot (FeatureScatter) for feature-feature relationships
plot1 <- FeatureScatter(pbmcdf, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmcdf, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```
### Data Normalization
Subset data for further analysis. The "LogNormalize" method can normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result.

```{r, warning=FALSE, message=FALSE}
#pbmcdf <- NormalizeData(pbmcdf, normalization.method = "LogNormalize", scale.factor = 10000)
# Default
pbmcdf <- NormalizeData(pbmcdf)

```
### Feature selection
Calculate a subset of features which are highly expressed in some cells, and lowly expressed in others
```{r}
pbmcdf <- FindVariableFeatures(pbmcdf, selection.method = "vst", nfeatures = 2000)
# top 5 most highly variable genes
top5 <- head(VariableFeatures(pbmcdf), 5)
# plot features
plot1 <- VariableFeaturePlot(pbmcdf)
plot2 <- LabelPoints(plot = plot1, points = top5, repel = TRUE)
plot1 + plot2

```
### Data scaling
A linear transformation ('scaling') is processed prior to dimensional reduction techniques, PCA.
```{r, warning=FALSE, message=FALSE}
pbmcdf <- ScaleData(pbmcdf)
# Remove unwanted sources of variation
pbmcdf <- ScaleData(pbmcdf, vars.to.regress = "percent.mt")

```
### Dimensional reduction_PCA
Linear dimensional reduction

```{r, warning=FALSE, message=FALSE}
# PCA
pbmcdf <- RunPCA(pbmcdf, features = VariableFeatures(object = pbmcdf))
print(pbmcdf[["pca"]], dims = 1:10, nfeatures = 10)
# Visualization
VizDimLoadings(pbmcdf, dims = 1:2, reduction = "pca")
DimPlot(pbmcdf, reduction = "pca")
DimHeatmap(pbmcdf, dims = 1, cells = 1000, balanced = TRUE)
DimHeatmap(pbmcdf, dims = 1:5, cells = 1000, balanced = TRUE)

```
### Select componenets
Identify 'significant' PCs 
```{r}
pbmcdf <- JackStraw(pbmcdf, num.replicate = 100)
pbmcdf <- ScoreJackStraw(pbmcdf, dims = 1:10)
JackStrawPlot(pbmcdf, dims = 1:10)
ElbowPlot(pbmcdf)

```

### Cell Clustering
First construct a KNN graph based on the euclidean distance in PCA space, next apply modularity optimization techniques (Louvain algorithm (default) or SLM) to iteratively group cells together. Clusters can be found by  Idents function.

```{r}
# Cluster
pbmcdf <- FindNeighbors(pbmcdf, dims = 1:10)
pbmcdf <- FindClusters(pbmcdf, resolution = 0.5)
#cluster IDs of top 20 cells
head(Idents(pbmcdf), 20)
```
### Dimensional reduction_UMAP

Non-linear dimensional reduction

```{r}
pbmcdf <- RunUMAP(pbmcdf, dims = 1:10)
DimPlot(pbmcdf, reduction = "umap")
saveRDS(pbmcdf, file = "output/pbmcdf_tutorial.rds")

```
### Differentially expressed features

```{r}

# find all markers of cluster 3
cluster3.markers <- FindMarkers(pbmcdf, ident.1 = 3, min.pct = 0.25)
head(cluster3.markers, n = 5)
# find all markers distinguishing cluster 4 from clusters 0 and 3
cluster4.markers <- FindMarkers(pbmcdf, ident.1 = 4, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster4.markers, n = 5)

# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmcdf.markers <- FindAllMarkers(pbmcdf, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmcdf.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)

cluster3.markers <- FindMarkers(pbmcdf, ident.1 = 3, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)

VlnPlot(pbmcdf, features = c("LINC00926", "TCL1A"))

# you can plot raw counts as well
VlnPlot(pbmcdf, features = c("CCL5", "GZMK"), slot = "counts", log = TRUE)

FeaturePlot(pbmcdf, features = c("GZMA", "CD2", "AQP3", "CD14", "S100A8", "CST7"))
top5 <- pbmcdf.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)
DoHeatmap(pbmcdf, features = top5$gene) + NoLegend()

```



### Assigning cell type to clusters
Use canonical markers to match clustering to cell types
```{r}
new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T", "FCGR3A+ Mono", 
                     "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmcdf)
pbmcdf <- RenameIdents(pbmcdf, new.cluster.ids)
DimPlot(pbmcdf, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
#saveRDS(pbmcdf, file = "output/pbmcdf3k_final.rds")

```
